### PR TITLE
efabless ring IO template + hooks

### PR DESCRIPTION
**Pull request by harrisonliew**
This allows us to make a customized IO ring in Innovus that matches exactly what efabless will accept.

@ethanwu10 @ezhes @ansaschmulbach

**Related PRs / Issues**

**Type of change**:
- [ ] Bug fix
- [X] New feature
- [ ] Other enhancement

**Impact**:
- [ ] Change to core Hammer
- [X] Change to a Hammer plugin
- [ ] Other

**Contributor Checklist**:
- [X] Did you set `master` as the base branch?
- [X] Did you state the type-of-change/impact?
- [X] Did you delete any extraneous prints/debugging code?
- [X] (If applicable) Did you add documentation for the feature?
- [ ] (If applicable) Did you update the `poetry.lock` file if you updated the requirements in `pyproject.toml`?
- [ ] (If applicable) Did you add a unit test demonstrating the PR?
- [ ] (If applicable) Did you run this through the e2e integration tests?
- [ ] (If applicable) Did you update the submodules in `e2e/` if this feature depends on updated plugins?

### CODE REVIEW

File: hammer/technology/sky130/README.md [hunk 1]
```diff
@@ -63,7 +63,40 @@ We use the Calibre decks in the ``s8`` PDK, version ``V2.0.1``,
...
 
+### Sky130 s8io Library
 
+The IO ring expected by efabless for MPW/ChipIgnite can be created in Innovus using the NDA s8iom0s8 IO cell library. Here are the steps to use it:
+
+1. `extra/efabless_template.io` is a template IO file. You should modify this by replacing the `<inst_path>`s with the paths to the IO cell instances in your netlist. **DO NOT MODIFY THE POSITIONS OF THE CELLS OR REPLACE CLAMP CELLS WITH IO CELLS**
+
+    a. The ordering in the instance lists are from left to right (for top/bottom edges) and **bottom to top (for left/right edges)**. 
+
+    b. Refer [this documentation](https://skywater-pdk.readthedocs.io/en/main/contents/libraries/sky130_fd_io/docs/user_guide.html) for how to configure the pins of the GPIOs (not exhaustive).
```
Comment (rahulk29) [inline hammer/technology/sky130/README.md:74]: nit: Refer **to**

File: hammer/technology/sky130/README.md [hunk 2]
```diff
@@ -63,7 +63,40 @@ We use the Calibre decks in the ``s8`` PDK, version ``V2.0.1``,
...
+
+    c. By default, the `s8iom0s8_top_gpiov2` cell has a `tie_hi_esd` pin that must be tied to VDDIO, which is difficult to do in Innovus. It is suggested to make a custom version of this with that pin tied to VDDIO, export the GDS, and hack the CDL/GDS/LEF for that cell only, then add is separately as an extra library.
```
Comment (rahulk29) [inline hammer/technology/sky130/README.md:76]: nit: add is separately -> add **it** separately

File: hammer/technology/sky130/README.md [hunk 3]
```diff
@@ -63,7 +63,40 @@ We use the Calibre decks in the ``s8`` PDK, version ``V2.0.1``,
...
+        y: 0
+        width: 3588
+        height: 5188
+        margins:
+          left: 200.1
+          right: 200.1
+          top: 201.28
+          bottom: 201.28
+
+3. In your CLIDriver, you must import the following hook from the tech plugin and insert is as a `post_insertion_hook` after `floorplan_design`.
```
Comment (rahulk29) [inline hammer/technology/sky130/README.md:95]: nit: insert is -> insert **it**

**Review by nayiri-k — APPROVED** [review]

File: hammer/technology/sky130/__init__.py [hunk 4]
```diff
@@ -160,6 +161,32 @@ def setup_lvs_deck(self) -> None:
...
+        source_path = sky130A_path / 'libs.ref' / 'sky130_fd_io' / 'lef' / 'sky130_ef_io.lef'
+        if not source_path.exists():
+            raise FileNotFoundError(f"IO LEF not found: {source_path}")
+
+        cache_tech_dir_path = Path(self.cache_dir)
+        os.makedirs(cache_tech_dir_path, exist_ok=True)
+        dest_path = cache_tech_dir_path / 'sky130_ef_io.lef'
+
+        with open(source_path, 'r') as sf:
+            with open(dest_path, 'w') as df:
```
Comment (rahulk29) [inline hammer/technology/sky130/__init__.py:176]: nit: can reduce the level of nesting by one:
```
with open(source_path, 'r') as sf, open(dest_path, 'w') as df:
```

**Review by rahulk29 — APPROVED** [review]

**Review by nayiri-k — APPROVED** [review]

**Comment by ethanwu10** [comment]
(setting the PR back to draft until I actually finish a test run of the new changes...)

**Comment by ethanwu10** [comment]
I have [a script](https://github.com/ucb-bar/stac-top/blob/stac-master/vlsi/scripts/gen-io-file.py) for generating the IO file given a YAML description of the pin map, but I'll work on upstreaming it into hammer in a separate PR so we can get this merged first for the Chipyard release. Mostly, it just needs cleaning up to disentangle IO file generation from the mapping of signal names to instance names used by the chip's custom IOBinder implementation.

**Comment by harrisonliew** [comment]
Thanks @ethanwu10 for finishing this off!

**Pull request merged**